pkg/mcp/server/server.go

```diff
@@ -99,11 +125,84 @@ type AuthChecker interface {
 	Check(authInfo credentials.AuthInfo) error
 }
 
+type mailboxState int
+
+const (
+	mailboxStateReady mailboxState = iota
+	mailboxStateClosed
+)
+
 // watch maintains local state of the most recent watch per-type.
 type watch struct {
-	cancel       func()
-	nonce        string
-	responseChan chan *WatchResponse
+	// only accessed from connection goroutine
+	cancel          func()
+	nonce           string // most recent nonce
+	nonceVersionMap map[string]string
+
+	// NOTE: do not hold `mu` when reading/writing to this channel
+	mailboxReadyChan chan mailboxState
+
+	mu                      sync.Mutex
+	mailbox                 *WatchResponse
+	mostRecentNackedVersion string
+	timerRunning            bool
+	timer                   *time.Timer
+}
+
+func (w *watch) schedulePush() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.mailbox == nil {
+		return
+	}
+
+	retry := false
+	if w.mostRecentNackedVersion != "" && w.mailbox.Version == w.mostRecentNackedVersion {
+		if !w.timerRunning {
+			w.timerRunning = true
+			w.timer = time.AfterFunc(nackLimitFreq, func() {
+				w.mu.Lock()
+				w.timerRunning = false
+				w.mu.Unlock()
+
+				select {
+				case w.mailboxReadyChan <- mailboxStateReady:
+				default:
+					retry = true
+				}
+			})
+		}
+	} else {
+		if w.timerRunning {
+			w.timerRunning = false
+			if w.timer.Stop() {
+				<-w.timer.C
+			}
+		}
+		select {
+		case w.mailboxReadyChan <- mailboxStateReady:
+		default:
+			retry = true
+		}
+	}
+
+	if retry {
+		time.AfterFunc(retryPushDelay, w.schedulePush)
+	}
+}
+
+func (w *watch) pushResponse(response *WatchResponse) {
+	w.mu.Lock()
+	w.mailbox = response
+	w.mu.Unlock()
+
+	if response == nil {
```

Comment:
It looks like this can be handled in schedulePush

Reply:
Renamed and documented the functions. I think we need two functions here: a higher-level saveAndPush function is that is called once per-desired push. This function also handles watch shutdown. The second lower-level function is responsible for the actual push and reschedules itself as necessary in case the serving layer is busy.

Reply:
We need to differentiate the user closing the watch vs. an already saved response being picked up and sent by the lower-level function. Alternatively, we could add a new boolean parameter to `server.PushResponseFunc`/`saveResponseAndSchedulePush` to indicate close, i.e.

```golang
func saveResponseAndSchedulePush(response *WatchResponse, close bool) {
	w.mu.Lock()
    defer w.mu.Unlock()

    if close {
        w.newPushResponse = nil
        w.newPushResponseReadyChan <- newPushResponseStateClosed
     	return
    }

    w.newPushResponse = response
    w.schedulePush()
}
```

Reply:
_refactored_: I kept the Watch() method signature but moved the close handling into the schedulePush function.